src/nvim/os/stream.c

```diff
@@ -0,0 +1,27 @@
+// Functions for working with stdio streams (as opposed to RStream/WStream).
+
+#include <stdio.h>
+#include <stdbool.h>
+
+#include <uv.h>
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/stream.c.generated.h"
+#endif
+
+/// Sets the stream associated with `fd` to "blocking" mode.
+///
+/// @return `0` on success, or `-errno` on failure.
+int stream_set_blocking(int fd, bool blocking)
+{
+  int flags = fcntl(fd, F_GETFL, 0);
+  int err = 0;
+  if (!blocking && !(flags & O_NONBLOCK)) {
+    err = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+  }
+  else if (blocking && (flags & O_NONBLOCK)) {
+    err = fcntl(fd, F_SETFL, flags & ~O_NONBLOCK);
+  }
```

Comment:
I'd much prefer to use the libuv-based solution, but this is a temporary bandaid to fix the bug. There's a remote possibility that it's a libuv bug: https://github.com/joyent/libuv/issues/838#issuecomment-58621040 

Or it's related to use of `uv_poll_t` (doesn't seem to be the case for us). https://github.com/yfinkelstein/node-zookeeper/pull/101